### PR TITLE
Adding debug information for time usage in wptrunner

### PR DESCRIPTION
**Pull request by WeizhongX**
We print out the time used by each sub processes to understand if the
work load is evenly distributed among the sub processes.

We use --debug-runner-logging to enable such logging.

### CODE REVIEW

**Comment by WeizhongX** [comment]
Folks, I pushed this patch for testing purpose. The one with the command line argument will be pushed later.

From the test here, I did see the workload among workers are unbalanced.

 7:23.30 INFO Test timing for testharness tests:
 7:23.30 INFO   350.25 total testing time
 7:23.30 INFO Thread timing:
 7:23.30 INFO       worker/25:     88 tests, 280.15 secs
 7:23.30 INFO       worker/30:     68 tests, 238.17 secs
 7:23.30 INFO       worker/33:     78 tests, 276.54 secs
 7:23.30 INFO       worker/31:     77 tests, 349.91 secs
 7:23.30 INFO       worker/36:     84 tests, 271.78 secs
 7:23.30 INFO       worker/40:     87 tests, 320.43 secs
 7:23.30 INFO       worker/23:     78 tests, 288.63 secs
 7:23.30 INFO       worker/38:     84 tests, 290.50 secs
 7:23.30 INFO       worker/26:     78 tests, 286.00 secs
 7:23.30 INFO       worker/27:    100 tests, 315.42 secs
 7:23.30 INFO       worker/34:     81 tests, 302.77 secs
 7:23.30 INFO       worker/32:     75 tests, 258.61 secs
 7:23.30 INFO       worker/37:     84 tests, 285.50 secs
 7:23.30 INFO       worker/39:     70 tests, 237.37 secs
 7:23.30 INFO       worker/35:     78 tests, 244.77 secs
 7:23.30 INFO       worker/24:     91 tests, 269.16 secs
 7:23.30 INFO       worker/29:     80 tests, 290.91 secs
 7:23.30 INFO       worker/28:     79 tests, 250.26 secs
 7:23.30 INFO       worker/22:     58 tests, 185.55 secs
 7:23.30 INFO       worker/21:     77 tests, 259.59 secs
 7:23.30 INFO     5502.02 cumulative, 275.10 optimal

**Comment by jgraham** [comment]
So for pure testing patches, it's a good idea to open the PR as a draft to avoid unnecessary notifications.

That said, some feedback:
* We already have some similar slightly more generic infrastructure with the "instruments" feature. See instruments.py for the implementation and use `--instrument-to-file` to enable it. It doesn't explicitly provide the exact data you have here, but does provide enough data to construct the information you want.
* In general there's a tradeoff between exactly balancing different processes to minimise runtime, and having predictable test execution order. Traditionally we have favoured predictability over performance (e.g. using by-dir chunking in prod so that subdirectories always run in the same process). On the other end of the scale there could be a single queue for all tests, with each process pulling a test off the queue. That would even up the runtimes, but any interaction between tests would lead to flaky results. Early on I tried some more complex heuristics to try and balance chunks (e.g. by estimating the maximum runtime of each chunk from the timeout values of each test, and then trying to equalise the maximum possible runtime in each chunk), but that was hard to understand and didn't produce especially good results. So I guess the question is what are you trying to achieve, and what tradeoffs are you prepared to make?

**Comment by WeizhongX** [comment]
jgraham@, thanks for the feedback. The goal here is to make wptrunner run faster, so that we can use it in CQ, where test will be triggered each time there is a new change.

I do not have much data for now, so can not say what tradeoffs I prepared to make. I definitely don't want to increase flakiness a lot. (I am thinking using "by-dir" might be fine, but we might need separate slow/fast tests in each directories, and schedule the slow tests first?)

About flakiness, I am not sure what is the flaky rate for now. And I did not see we retried failed tests. Do we need add such mechanism? (In chromium we are retrying the failed tests up to 3 times, and the test is considered pass if it passed in any retry, to tolerate the flakiness.)

**Comment by WeizhongX** [comment]
jgraham@, for the goal to make wptrunner run faster, It would be very appreciated if you have any suggestion on this.

**Comment by WeizhongX** [comment]
Log from another run. I see we got 0 tests on worker/30, and slowest v.s. fatest is 465.88 v.s. 209.18. There is a bug here?

 9:41.24 INFO Test timing for testharness tests:
 9:41.24 INFO   466.46 total testing time
 9:41.24 INFO Thread timing:
 9:41.24 INFO       worker/30:      0 tests, 0.00 secs
 9:41.24 INFO       worker/22:     59 tests, 209.18 secs
 9:41.24 INFO       worker/34:     80 tests, 235.52 secs
 9:41.24 INFO       worker/26:     66 tests, 243.94 secs
 9:41.24 INFO       worker/31:     62 tests, 248.24 secs
 9:41.24 INFO       worker/25:     85 tests, 251.64 secs
 9:41.24 INFO       worker/39:     77 tests, 251.92 secs
 9:41.24 INFO       worker/23:     82 tests, 254.13 secs
 9:41.24 INFO       worker/33:    100 tests, 256.56 secs
 9:41.24 INFO       worker/38:     86 tests, 265.64 secs
 9:41.24 INFO       worker/24:     83 tests, 273.66 secs
 9:41.24 INFO       worker/29:     71 tests, 277.85 secs
 9:41.24 INFO       worker/28:     69 tests, 280.33 secs
 9:41.24 INFO       worker/32:     80 tests, 312.47 secs
 9:41.24 INFO       worker/35:     87 tests, 313.72 secs
 9:41.24 INFO       worker/40:     86 tests, 326.78 secs
 9:41.24 INFO       worker/27:     77 tests, 334.38 secs
 9:41.24 INFO       worker/37:    105 tests, 344.96 secs
 9:41.24 INFO       worker/21:     79 tests, 353.54 secs
 9:41.24 INFO       worker/36:    161 tests, 465.88 secs
 9:41.24 INFO     5500.33 cumulative, 275.02 optimal

**Comment by jgraham** [comment]
We don't currently have any max-retries mechanism. I can imagine being broadly supportive of such a thing, but it would need an RFC (for example I can imagine using it in Gecko CI if we could annotate the specific tests that require retries rather than it just being a global setting. But it sounds like you want a global setting).

Whether there's a bug with the parallelisation depends on exactly what you're doing. Tests are divided up into groups using the code at https://github.com/web-platform-tests/wpt/blob/master/tools/wptrunner/wptrunner/testloader.py#L362-L508. Depending on the number of tests you're running and the configuration, it might be that you don't end up with enough groups, so one worker ends up doing nothing. Or it might, as you say, be a bug. It would be pretty helpful if you add some diagnosics to figure out how the tests are being divided up.

Note: this is not an area where I strongly feel the current code is "right". If you want to make design changes, that might be acceptable, but not breaking existing CI systems is important, so there are constraints. In particular Gecko these days is passing in an explicit list of tests that are in each group (`GroupFIleTestSource` and no chunking). That means we can put test selection and grouping logic outside wptrunner. But the GitHub CI is still relying on the wptrunner code for splitting the tests into chunks and then forming groups within those chunks.

Let me know if you want more information here; I understand this stuff has quite a lot of legacy complexity.

File: tools/wptrunner/wptrunner/testrunner.py [hunk 1]
```diff
@@ -918,6 +923,16 @@ def run(self, test_type, tests):
             self.pool.add(manager)
         self.wait()
 
+        self.logger.info("Test timing for %s tests:" % test_type)
```
[the remark this answers is not in the document]
Comment (WeizhongX) [reply tools/wptrunner/wptrunner/testrunner.py:926]: I can change this to use logger.debug. jgraham@ previously mentioned instruments, I will need to check how to use that facility.

Thanks for all the other advices. I will take time to look into that.

**Comment by WeizhongX** [comment]
Sorry for the delay. I managed to put some time on this today. I am thinking the major problem is in SingleTestSource. (This is the class we used when running wptrunner in chromium.) I see this class divides tests into groups of the same number of runners. This will make each group has too many tests, then easily cause time distribution between workers are not even. I can do an experiment here to divide tests into smaller groups to see if that helps.

Some other thoughts that could be less effective and may need more effort:
I see we are running tests in rounds based on test types(crashtest, print-reftest, testharness, reftest, wdspec). For each round we will have some time wasted waiting for other runners to finish. If we can combine those to just one round, we will be able to get rid of some waiting. We may need to move some logic to TestRunner so that it can load correct classes.

I see  we have a TestRunnerManager class. It has a state machine and runs in its own thread. And it will try to restart the TestRunner process each time when there is a new group of tests. Do we need restart TestRunner for such cases? I understand we need restart the browser from time to time, to make sure it has a clean state.

**Comment by WeizhongX** [comment]
>> On the other hand, not grouping in any way seems potentially bad for reliability, since the tests end up running in a random order. So I wonder if that's what you'll want in the long run.
I think I will group the test cases by the directories they were in.  (We have another runner running wpt/web tests. I was very much copying what it did there.)

>> Juast being able to kill the runner process and have the OS ensure that we've cleaned up all the resources it was using makes it easier to reason about the state that we're in.
My feeling is that if the runner process is in a wrong state, it will affect all test cases? Then we will notice that?

We did not pass `--no-restart-on-unexpected`. We did have some sort of expectation data, but that is derived from somewhere else. So I would not say the quality of expectation data is super good.

**Comment by WeizhongX** [comment]
I tried to pass in '--run-by-dir=0' to run the test. With this, the work load does looks evenly distributed among workers, except there is still one thread got 0 test.

10:07.60 INFO Test timing for testharness tests:
10:07.60 INFO   462.11 total testing time
10:07.60 INFO Thread timing:
10:07.60 INFO       worker/ 2:      0 tests, 0.00 secs
10:07.60 INFO       worker/ 3:     60 tests, 356.43 secs
10:07.60 INFO       worker/16:     98 tests, 393.37 secs
10:07.60 INFO       worker/ 1:     52 tests, 394.72 secs
10:07.60 INFO       worker/ 9:     89 tests, 394.74 secs
10:07.60 INFO       worker/12:     86 tests, 395.71 secs
10:07.60 INFO       worker/ 6:    107 tests, 396.10 secs
10:07.60 INFO       worker/11:     79 tests, 399.82 secs
10:07.60 INFO       worker/ 8:     88 tests, 400.19 secs
10:07.60 INFO       worker/17:     90 tests, 400.79 secs
10:07.60 INFO       worker/ 0:     76 tests, 401.04 secs
10:07.60 INFO       worker/13:     98 tests, 402.78 secs
10:07.60 INFO       worker/10:     64 tests, 404.11 secs
10:07.60 INFO       worker/ 5:     78 tests, 404.65 secs
10:07.60 INFO       worker/ 4:     86 tests, 405.06 secs
10:07.60 INFO       worker/ 7:    111 tests, 408.17 secs
10:07.60 INFO       worker/15:     65 tests, 410.06 secs
10:07.60 INFO       worker/19:     81 tests, 417.86 secs
10:07.60 INFO       worker/18:    101 tests, 438.88 secs
10:07.60 INFO       worker/14:     95 tests, 461.72 secs
10:07.60 INFO     7686.22 cumulative, 384.31 optimal

But the problem is that this does not speed up the test. With 15 shards, the total time used on all shards is 182 minutes. Previously it is about 140 minutes.

Search "INFO Starting runner" in the log I got 947 times of such log.

I am trying to add some more instrumental information to understand the cost of starting the runner. Some preliminary results below shows about 3.5 seconds each time? I will upload a patch later.

MainThread;startup 36
MainThread;startup;load_tests 7097
MainThread;startup;start_environment 4358
MainThread;startup;ensure_environment 441
MainThread;startup 1
TestRunnerManager-testharness-0;testrunner;startup 0
TestRunnerManager-testharness-0;testrunner;start_init 0
TestRunnerManager-testharness-0;testrunner;init 513
TestRunnerManager-testharness-0;testrunner;start_runner 3480
TestRunnerManager-testharness-0;testrunner;test;FileAPI;Blob-methods-from-detached-frame.html 1032
TestRunnerManager-testharness-0;testrunner;after-test 0
TestRunnerManager-testharness-0;testrunner;restart_runner 0
TestRunnerManager-testharness-0;testrunner;stop_runner 43
TestRunnerManager-testharness-0;testrunner;init 512
TestRunnerManager-testharness-0;testrunner;start_runner 3493
TestRunnerManager-testharness-0;testrunner;test;FileAPI;FileReader;progress_event_bubbles_cancelable.html 792
TestRunnerManager-testharness-0;testrunner;after-test 0
TestRunnerManager-testharness-0;testrunner;test;FileAPI;FileReader;workers.html 636
TestRunnerManager-testharness-0;testrunner;after-test 0
TestRunnerManager-testharness-0;testrunner;restart_runner 0
TestRunnerManager-testharness-0;testrunner;stop_runner 56
TestRunnerManager-testharness-0;testrunner;init 515
TestRunnerManager-testharness-0;testrunner;start_runner 3446
TestRunnerManager-testharness-0;testrunner;test;FileAPI;FileReaderSync.worker.html 957
TestRunnerManager-testharness-0;testrunner;after-test 0
TestRunnerManager-testharness-0;testrunner;restart_runner 0
TestRunnerManager-testharness-0;testrunner;stop_runner 47
TestRunnerManager-testharness-0;testrunner;init 512
TestRunnerManager-testharness-0;testrunner;start_runner 3465
TestRunnerManager-testharness-0;testrunner;test;FileAPI;blob;Blob-array-buffer.any.html 847
TestRunnerManager-testharness-0;testrunner;after-test 0
TestRunnerManager-testharness-0;testrunner;test;FileAPI;blob;Blob-array-buffer.any.worker.html 752
TestRunnerManager-testharness-0;testrunner;after-test 0
TestRunnerManager-testharness-0;testrunner;test;FileAPI;blob;Blob-constructor-dom.window.html 654
TestRunnerManager-testharness-0;testrunner;after-test 0

**Comment by WeizhongX** [comment]
@jgraham, is there a way we can also instrument the child processes?

**Comment by WeizhongX** [comment]
I tried to collect more data to see how time is spent when restarting TestRunner. Previously I said that restarting TestRunner takes about 3.5 seconds. Now from the instrument.txt it shows about 0.5 seconds to restart chromedriver, and 3 seconds to restart chrome and connect to it. (I added more instrument code in BrowserManager.init, and WebDriverProtocol.connect when entering the leaving the two functions.) start_chrome.txt is a verbose log for starting chrome and connecting to it. From there I see starting chrome itself takes about 2 seconds. We are taking action to see if we can do anything about this, and how much time that can save us.

**Comment by WeizhongX** [comment]
[start_chrome.txt](https://github.com/web-platform-tests/wpt/files/8103295/start_chrome.txt)
[instrument.txt](https://github.com/web-platform-tests/wpt/files/8103296/instrument.txt)

**Pull request closed without merging**